### PR TITLE
Bug 1790046 - Add a nightly task to update FOG

### PR DESCRIPTION
**Pull request by badboy**
This requires changes in m-c before we can merge that here.
This is an import of https://github.com/mozilla/fog-update-bot, because it turns out keeping the action within this repository is actually easier.

### CODE REVIEW

File: .github/workflows/update-fog.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+name: "Update FOG"
+
+permissions:
+  contents: write
+  pull-requests: write
+
+on:
+  schedule:
+    # 04:00 UTC - every morning
```
Comment (chutten) [inline .github/workflows/update-fog.yml:17]: In what way are we guaranteed that this is interpreted wrt UTC?

File: fog-updater/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+and creates a new Pull Request against `probe-scraper` if it contains any changes.
+
+## Environment variables
+
+| Name | Description |
+| ---- | ----------- |
+| `DEBUG` | If set enables debug logging |
+| `DRY_RUN` | If set to `True` will not create a PR |
+| `GITHUB_REPOSITORY_OWNER` | The owner of the `probe-scraper` repository |
+| `AUTHOR_NAME` | The name name to use for the commit |
```
Comment (chutten) [inline fog-updater/README.md:15]:
```suggestion
| `AUTHOR_NAME` | The name to use for the commit |
```
The Department of Redundancy Department would like a word.

File: fog-updater/src/fog_update.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+# file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+from github import Github, GithubException, InputGitAuthor, enable_console_debug_logging
+import datetime
+import io
+import os
+import requests
+import sys
+
+DEFAULT_ORGANIZATION = "badboy"
```
Comment (chutten) [inline fog-updater/src/fog_update.py:14]: Should this be `mozilla` now that we're here?

File: fog-updater/src/fog_update.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+DEFAULT_AUTHOR_NAME = "data-updater"
+DEFAULT_AUTHOR_EMAIL = "jrediger@mozilla.com"
```
Comment (chutten) [inline fog-updater/src/fog_update.py:16]: Could use `telemetry-alerts@` if we wanted to grant that weird pseudo-address its own account, I guess.

Depends on how much you'd like your email to not be in code.

File: fog-updater/src/fog_update.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+USAGE = "usage: fog-update"
+REVIEWERS = ["@mozilla/glean"]
+INDEX_URL = "https://raw.githubusercontent.com/mozilla/gecko-dev/master/toolkit/components/glean/metrics_index.py"  # noqa
+BODY_TEMPLATE = """
+This (automated) patch updates the list from metrics_index.py.
+
+For reviewers:
+
+* Canonical source for the index: <https://searchfox.org/mozilla-central/source/toolkit/components/glean/metrics_index.py>
```
Comment (chutten) [inline fog-updater/src/fog_update.py:25]: Didn't I have a comment about making sure this and `INDEX_URL` match?

File: fog-updater/src/fog_update.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+* Please double-check that the changes here are valid and that the referenced files exist.
```
Comment (chutten) [inline fog-updater/src/fog_update.py:26]: It's good to include consequences. "If the referenced files do not exist, schema deploys will fail" (though I suppose the CI run on the PR will fail first?

File: fog-updater/src/fog_update.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+* Please double-check that none of the files are _deleted_ from the list.
```
Comment (chutten) [inline fog-updater/src/fog_update.py:27]: "Deleted files result in incompatible schema changes that might not be immediately caught."

File: fog-updater/src/fog_update.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+* Delete this branch after merging or closing the PR.
+
+---
+
+The source code of this automation bot lives in [badboy/fog-update-bot](https://github.com/badboy/fog-update-bot).
```
Comment (chutten) [inline fog-updater/src/fog_update.py:32]: Not any more it doesn't!

File: fog-updater/src/fog_update.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+
+def eval_extract(code):
+    """
+    Eval `code` and return a map of variables and their values.
+
+    `code` should be valid Python code.
+    Only the builtins `list` and `set` are provided.
+
+    Note: this execute arbitrary Python code.
```
Comment (chutten) [inline fog-updater/src/fog_update.py:47]:
```suggestion
    Note: this executes arbitrary Python code.
```

File: fog-updater/src/fog_update.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+
+    new_content = content
+    for item in data:
+        name, metrics_or_pings, library, files = item
+        new_content = swap_file_list(
+            new_content, name, files, metrics_or_pings, library
+        )
+
+    if content == new_content:
+        raise Exception(
```
Comment (chutten) [inline fog-updater/src/fog_update.py:128]: The Software Engineer in me objects to using exceptions to mark unexceptional (common/normal) situations.

The pragmatic programmer in me doesn't see that it makes any difference, though.

File: fog-updater/src/fog_update.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    except Exception as e:
+        print(f"{ts()} {e}")
+        return
+
+    if dry_run:
+        print(f"{ts()} Dry-run so not continuing.")
+        return
+
+    # Create a non unique PR branch name for work on this ac release branch.
+    pr_branch_name = f"fog-update/update-metrics-index-{short_version}"
```
Comment (chutten) [inline fog-updater/src/fog_update.py:189]: TIL branch names can contain `/`

File: fog-updater/src/fog_update.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+    try:
+        pr_branch = repo.get_branch(pr_branch_name)
+        if pr_branch:
+            print(f"{ts()} The PR branch {pr_branch_name} already exists. Exiting.")
```
Comment (chutten) [inline fog-updater/src/fog_update.py:194]: So this means if we aren't quick enough to review these each day, the work will pile up?

File: fog-updater/src/fog_update.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+            return
+    except GithubException:
+        # TODO Only ignore a 404 here, fail on others
```
Comment (chutten) [inline fog-updater/src/fog_update.py:197]: 👀

File: fog-updater/src/test_util.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    "F"
+]
+"""
+
+
+def test_eval_metrics_index():
+    data = eval_extract(METRICS_INDEX)
+    assert data["first_yamls"] == ["A", "B"]
+    assert data["second_yamls"] == ["B", "C"]
+    assert data["metrics_yamls"] == ["A", "B", "C"]
```
Comment (chutten) [inline fog-updater/src/test_util.py:45]: Should probably also assert something about pings_yamls

File: fog-updater/src/test_util.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+from fog_update import eval_extract, swap_file_list
+
+
+REPOSITORIES_YAML = """
+---
+version: "2"
+applications:
+  - app_name: firefox_desktop
+    metrics_files: # When adding here, consider if you should also add to pine.
```
Comment (chutten) [inline fog-updater/src/test_util.py:14]: Are you keeping this comment because it's hilarious?

File: fog-updater/src/test_util.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+    # New files added.
+    assert "- METRICS_FILES" in output
+    assert "- A" not in output
+    # ping files untouched.
+    assert "- PING_FILES" in output
+
+    # Other app untouched
+    assert "- OTHER_METRICS_FILES" in output
+    assert "- OTHER_PING_FILES" in output
```
Comment (chutten) [inline fog-updater/src/test_util.py:82]: Wouldn't mind a test covering `ping_files`

**Review by chutten — CHANGES_REQUESTED** [review]
This seems like a thing that should be linked to from documentation. Maybe in FOG's docs we should link directly to this dir? Otherwise it might become some of that semi-hidden magic that mostly works and only like two people know exists.

File: .github/workflows/update-fog.yml (hunk 1 above)
Comment (badboy) [reply .github/workflows/update-fog.yml:17]: GitHub Actions docs [say so](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) (hm, might want to shift that by a bit just to not run into the called out "high load" periods.

File: fog-updater/src/fog_update.py (hunk 11 above)
Comment (badboy) [reply fog-updater/src/fog_update.py:189]: Whaaaaat? yes of course! 
You just can't have `fog-update` as a branch name then, at least in checkouts, as this is really just files in `.git/logs/refs/heads/`

File: fog-updater/src/fog_update.py (hunk 12 above)
Comment (badboy) [reply fog-updater/src/fog_update.py:194]: ... and we wouldn't even hear about it until we merge & delete and the next run comes around!

I'm ok with that for now.
We could switch to adding additional commits to the branch if it exists or always creating a fresh branch or something.

File: fog-updater/src/fog_update.py (hunk 13 above)
Comment (badboy) [reply fog-updater/src/fog_update.py:197]: aha copy'pasta from relbot.

File: fog-updater/src/test_util.py (hunk 15 above)
Comment (badboy) [reply fog-updater/src/test_util.py:14]: depends on who you ask! (Nah, I copied that from the original a while back in the before times)

**Comment by badboy** [comment]
Gecko work has been merged and https://firefox-source-docs.mozilla.org/toolkit/components/glean/user/new_definitions_file.html shows the current docs. We can land this now!

**Review by chutten — APPROVED** [review]

**Comment by badboy** [comment]
Let's merge and see! (Nothing will break because it's just a PR)

**Pull request merged**